tsdb/head_append.go

```diff
@@ -1389,17 +1393,84 @@ func (s *memSeries) appendPreprocessor(
 	return c, true, chunkCreated
 }
 
+// nativeHistogramsAppendPreprocessor takes care of cutting new native histogram chunks and m-mapping old ones. Native
+// histogram chunks are cut based on their size in bytes.
+// It is unsafe to call this concurrently with s.iterator(...) without holding the series lock.
+// This should be called only when appending data.
+func (s *memSeries) nativeHistogramsAppendPreprocessor(t int64, e chunkenc.Encoding, chunkDiskMapper *chunks.ChunkDiskMapper, chunkRange int64) (c *memChunk, sampleInOrder, chunkCreated bool) {
+	c = s.head()
+
+	if c == nil {
+		if len(s.mmappedChunks) > 0 && s.mmappedChunks[len(s.mmappedChunks)-1].maxTime >= t {
+			// Out of order sample. Sample timestamp is already in the mmapped chunks, so ignore it.
+			return c, false, false
+		}
+		// There is no head chunk in this series yet, create the first chunk for the sample.
+		c = s.cutNewHeadChunk(t, e, chunkDiskMapper, chunkRange)
+		chunkCreated = true
+	}
+
+	// Out of order sample.
+	if c.maxTime >= t {
+		return c, false, chunkCreated
+	}
+
+	if c.chunk.Encoding() != e {
+		// The chunk encoding expected by this append is different than the head chunk's
+		// encoding. So we cut a new chunk with the expected encoding.
+		c = s.cutNewHeadChunk(t, e, chunkDiskMapper, chunkRange)
+		chunkCreated = true
+	}
+
+	numSamples := c.chunk.NumSamples()
+	targetBytes := chunkenc.TargetBytesPerNativeHistogramChunk
+	numBytes := len(c.chunk.Bytes())
+
+	if numSamples == 0 {
+		// It could be the new chunk created after reading the chunk snapshot,
+		// hence we fix the minTime of the chunk here.
+		c.minTime = t
+		s.nextAt = rangeForTimestamp(c.minTime, chunkRange)
+	}
+
+	// If we reach 25% of a chunk's desired maximum size, predict an end time
+	// for this chunk that will try to make samples equally distributed within
+	// the remaining chunks in the current chunk range.
+	// At the latest it must happen at the timestamp set when the chunk was cut.
+	if !s.nativeHistogramChunkHasComputedEndTime && numBytes >= targetBytes/4 {
+		ratioToFull := float64(targetBytes) / float64(numBytes)
+		s.nextAt = computeChunkEndTime(c.minTime, c.maxTime, s.nextAt, ratioToFull)
+		s.nativeHistogramChunkHasComputedEndTime = true
+	}
+	// If numBytes > targetBytes*2 then our previous prediction was invalid. This could happen if the sample rate has
+	// increased or if the bucket/span count has increased.
+	// We also enforce a minimum sample count per chunk here.
+	// Note that next chunk will have its nextAt recalculated for the new rate.
+	nextChunkRangeStart := (chunkRange - (c.maxTime % chunkRange)) + c.maxTime
+	if (t >= s.nextAt || numBytes >= targetBytes*2) && (numSamples >= chunkenc.MinSamplesPerNativeHistogramChunk || t >= nextChunkRangeStart) {
```

Comment:
1. Why is it `targetBytes*2`? I would expect it to be capped at `targetBytes`.
2. Checking for `>= targetBytes` is a delayed indicator here, as in we already crossed our target, but we should try to stay under or equal to target (like how XOR chunk stays under 1024 bytes). In majority case, where clients or servers have limits on number of buckets, I would expect it to cross 10 samples easily and hit the size limit first. I will think about how this can be done efficiently (the brute force is to append and revert back if it crosses the target and cut a new chunk, but there could be a better way)

Reply:
For point 1, I took the idea from the original `appendPreprocessor`. My understanding is that `t >= s.nextAt` is the primary trigger to cut a new head chunk but `numBytes >= targetBytes * 2` is there as a backup trigger (if sample rate has increased or if bucket/span count has increased since `s.nextAt` was computed).

I suspect cutting a new head chunk when `numBytes >= targetBytes` might result in samples not being equally distributed within the chunks in the chunk range.

Reply:
Yeah, I agree.

Reply:
About the 2nd point: As said, we cannot ever give a hard limit for histogram chunks. A single histogram might take more than 1024 byte, or in other words: The worst case size of a single histogram sample is unlimited (while the worst case for a float sample is 19 bytes).

Therefore, we don't really have to worry about not always keeping the 1024 bytes. We have to handle the case where the size exceeds 1024 bytes anyway (and it has worked fine so far – Mimir has run tests with fairly random numbers in the histograms, creating chunks with 118 histograms and a total chunk size of ~70kiB).

So I would definitely not consider reverting.